### PR TITLE
Node10 fixes

### PR DESCRIPTION
**Pull request by aricart**
[FIX] #222 - socket upgrade used socket.end() which half closed the socket.
[FIX] Deprecations in node10
[FIX] fs.writeFile makes callback mandatory

### CODE REVIEW

File: lib/nats.js [hunk 1]
```diff
@@ -891,7 +890,6 @@ Client.prototype.processInbound = function() {
                             // if we have a stream, this is from an old connection, reap it
                             if (client.stream) {
                                 client.stream.removeAllListeners();
-                                client.stream.end();
```
Comment (kozlovic) [inline lib/nats.js]: You had introduced the call to `.end()` in some commits (https://github.com/nats-io/node-nats/commit/be8fa847de9691297f679d7ab7044411795fc543). Could you try recall what problem you tried to fix and that this is not going to re-introduce the issue? If there was a test added at the time and it still pass, I guess no.

File: lib/nats.js [hunk 2]
```diff
@@ -615,7 +615,7 @@ Client.prototype.createConnection = function() {
     // otherwise in addition to the leak events will fire fire
```
Comment (derekcollison) [inline lib/nats.js]: fire fire

File: benchmark/pub_perf.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ nats.on('connect', function () {
 
   var start = new Date();
 
-  var invalid2octet = new Buffer('\xc3\x28', 'binary');
+  var invalid2octet = Buffer.from('\xc3\x28', 'binary');
```
[the remark this answers is not in the document]
Comment (aricart) [reply benchmark/pub_perf.js]: @derekcollison We have not. We published a while back about maintaining support for node to match https://github.com/nodejs/Release. Why are they downloading the latest node-nats on an old version of node. Worse comes to worse we can always do a one-off.

File: lib/nats.js (hunk 2 above)
Comment (aricart) [reply lib/nats.js]: From looking at it, the stream cleanup code from:

```Client.prototype.closeStream = function() {
  if (this.stream !== null) {
    this.stream.end();
    this.stream.destroy();
```
was duplicated to the places were we could fail to close the stream.
 ` this.stream.removeAllListeners();` is the fix for #45 multiple notifications when listeners were left out.

File: lib/nats.js (hunk 1 above)
Comment (aricart) [reply lib/nats.js]: @kozlovic see below

File: benchmark/pub_perf.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aricart) [reply benchmark/pub_perf.js]: Buffer.from() works correctly on 4 and up.

[the remark this answers is not in the document]
Comment (StarpTech) [reply benchmark/pub_perf.js]: `Buffer.from` does not work on Node.js 4 but why is it relevant when Node.js 4 maintenance has reached end-of-life (04.30.2018).

File: lib/nats.js [hunk 4]
```diff
@@ -1219,7 +1217,7 @@ Client.prototype.publish = function(subject, msg, opt_reply, opt_callback) {
         }
         this.sendCommand(psub + Buffer.byteLength(str) + CR_LF + str + CR_LF);
     } else {
-        var b = new Buffer(psub.length + msg.length + (2 * CR_LF_LEN) + msg.length.toString().length);
+        var b = Buffer.alloc(psub.length + msg.length + (2 * CR_LF_LEN) + msg.length.toString().length);
```
Comment (StarpTech) [inline lib/nats.js]: Does it have a significant performance impact? The counterpart of `new Buffer` is `Buffer.allocUnsafe`

File: benchmark/pub_perf.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply benchmark/pub_perf.js]: It fails on node 4.0.0 but works on node 4.8.7

[the remark this answers is not in the document]
Comment (StarpTech) [reply benchmark/pub_perf.js]: New Buffer api was added in 4.5 https://nodejs.org/en/blog/release/v4.5.0/

[the remark this answers is not in the document]
Comment (aricart) [reply benchmark/pub_perf.js]: @StarpTech yes when we are talking about 4, we are talking about 4.x, which was until yesterday supported.

File: lib/nats.js (hunk 1 above)
Comment (kozlovic) [reply lib/nats.js]: Sorry, I missed the fact that we pass the socket as part of the options to `tls.Connect()`.

**Pull request merged**